boards: Add support for the HiFi4 DSP of the MIMXRT685

VitekST opened the pull request:
This PR implements a basic Zephyr port for the Cadence HiFi 4 DSP in the NXP MIMXRT685 microcontroller. Besides a barebones support it implements interrupt handling by driving the INPUTMUX peripheral, DSP code loading and run control from the ARM side of that microcontroller and basic peripherals (Flexcomm UART and GPIO) are also instantiated.

The changes are dependent on: https://github.com/zephyrproject-rtos/hal_nxp/pull/301

carlescufi commented:
> The changes are dependent on: [zephyrproject-rtos/hal_nxp#301](https://github.com/zephyrproject-rtos/hal_nxp/pull/301)

Can you point the manifest to the `pull/301/head` revision then please?

iuliana-prodan commented:
> Can you point the manifest to the `pull/301/head` revision then please?

Here's an example on how to do it: [commit 1](https://github.com/zephyrproject-rtos/zephyr/pull/62776/commits/fbdded5679a07499dfdf352369f263822546b569) from this PR: https://github.com/zephyrproject-rtos/zephyr/pull/62776/commits

iuliana-prodan commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,51 @@
+
+config SOC_TOOLCHAIN_NAME
+	string
+	default "nxp_rt6xx"
```
Is this toolchain available in zephyr-sdk?
I'm not aware of it.

Can we use the `nxp_rt500_adsp` the one used for RT585?

Have you compiled this support? How?
Probably with xcc, right?

VitekST replied:
Indeed it is not.
I haven't explored compilation using the said toolchain for the RT595, as I was instructed to just focus on the port itself by my supervisor. My guess is that it could cause some problems, as the ISAs are for sure different (HiFi 4 vs. Fusion F1).

I used `xt-clang`, as `xcc` is not even present in newer versions of the Cadence toolchain.

iuliana-prodan replied:
> Indeed it is not. I haven't explored compilation using the said toolchain for the RT595, as I was instructed to just focus on the port itself by my supervisor. My guess is that it could cause some problems, as the ISAs are for sure different (HiFi 4 vs. Fusion F1).
> 
Oh, you're right.

> I used `xt-clang`, as `xcc` is not even present in newer versions of the Cadence toolchain.

Most probably will need a toolchain for this target in zephyr sdk.
See examples: https://github.com/zephyrproject-rtos/sdk-ng/commit/7d1b95ecfa5b9ec45319eb4dda27dd6ce9bb94dc

dbaluta replied:
@VitekST  we will definitely need toolchain support in sdk-ng for your platform. As @iuliana-prodan  pointed the process can be easily done following an older platform addition. 

Mind that it will take some time until you merge the toolchain support and then we update the CI to use the new SDK. Expect 2-3 weeks on that if you follow up with maintainers once you submit your changes. 

We can help you with that, but you should just look at how we enabled rt595 or imx8ulp.

Without the toolchain support in the sdg-ng we cannot merge this PR as we need CI to compile it first.

danieldegrasse replied:
Just to be clear- since the imx8ulp has the same HiFi4 DSP, can we reuse the existing toolchain? Or will we need a new toolchain for this HiFi4 specifically?

VitekST replied:
@danieldegrasse I haven't studied the differences between the two instances of the DSP, but from what I've encountered, they can differ, as the Xtensa IP is a customizable one. Luckily, @iuliana-prodan told me over internal Teams that adapting GCC with related tools and libraries should be just a matter of collecting a few source files from Cadence's BSP, creating a new crosstool-ng overlay with it and compiling the toolchain.

VitekST replied:
@dbaluta @iuliana-prodan Toolchain PR created: https://github.com/zephyrproject-rtos/sdk-ng/pull/721

dbaluta commented:
@VitekST I would advise going through : https://cbea.ms/git-commit/#seven-rules

With a stress on:
* User imperative mood in the subject line https://cbea.ms/git-commit/#imperative
* wrap the commit body / commit subject at max 100 chars
* add some explanation for each commit with some info for the reviewer (like WHY the commit is needed, and what brings new to the existing code).

butok commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -44,7 +44,6 @@ struct gpio_mcux_lpc_config {
 	IOCON_Type *pinmux_base;
 #endif
 	uint32_t port_no;
-	clock_ip_name_t clock_ip_name;
```
Guess, you do not want to change LPC drivers in the RTxxx commit. Do you?

VitekST replied:
I do - I guess the driver is just named because of historical reasons. The `gpio_mcux_lpc.c` driver really is used on the RT685, despite it being an i.MX RT device and not an LPC device.

butok commented on drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.c:
```diff
@@ -0,0 +1,123 @@
+    uint32_t *dst = cfg->regions[section].base;
+    uint32_t size = image->length;
+
+    while(size > 0)
```
If to use "for" instead of "while", the code inside { } will have only one line.

butok commented on soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,348 @@
+/*
```
Copyright and License information is missed,

butok commented on soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2023 Google LLC.
+ # Copyright (c) 2023 NXP
```
# should be replaced by *

dbaluta replied:
Official NXP Copyright is : ```Copyright 2023 NXP```

iuliana-prodan commented on boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts:
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2023 Google LLC.
```
Why do you use the Google copyright?

It should be NXP, or if you used the rt5xx as example, you should add NXP copyright also.

VitekST replied:
That was done in a following commit.

iuliana-prodan replied:
yes, don't add a new commit to fix something from a commit that was not merged.

VitekST replied:
Commit history redone - no fixup commits.

iuliana-prodan commented on soc/xtensa/nxp_adsp/Kconfig.defconfig:
```diff
@@ -1,6 +1,6 @@
 # NXP i.MX8/RT SoC family default configuration options
 #
-# Copyright (c) 2021 NXP
+# Copyright (c) 2021,2023 NXP
```
s/Copyright (c) 2021,2023 NXP/Copyright 2021, 2023 NXP

dcpleung commented on boards/xtensa/nxp_adsp_rt685/Kconfig:
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) 2023 Google LLC.
+# Copyright (c) 2023 NXP
```
I think it is better to add to copyright when the files are first introduced, or first modified in the PR.

dcpleung commented on soc/xtensa/nxp_adsp/CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 # NXP i.MX/i.MX RT DSP family CMake file
 #
-# Copyright (c) 2021,2023 NXP
+# Copyright (c) 2021, 2023 NXP
```
This kind of changes should be squashed back to the commit introducing these lines originally.

dcpleung commented on include/zephyr/dt-bindings/misc/nxp-inputmux-rt6xx.h:
```diff
@@ -4,37 +4,8 @@
-#ifndef __NXP_ADSP_INPUTMUX_INTC_RT6XX_H__
-#define __NXP_ADSP_INPUTMUX_INTC_RT6XX_H__
-
-#define ADSP_IRQ_5 5
```
Could you find a way to unify this block of changes with previous commits? This is being introduced a few commits back but then are removed here, all in the same PR.

VitekST replied:
I redid the whole commit history, so no commits without comments and subsequent commits are there.

danieldegrasse reviewed (commented):
Thanks for this work. A general comment about commit formatting, somewhat echoing what @dbaluta said:

Commits should follow the following rules:
- focus on one area (SOC enablement, driver changes/new driver, new board, etc)
- each commit should build correctly
- fixup commits should be avoided. Squash changes in response to reviewers back into the original commit

From a broad view, I might suggest the following set of commits:
- changes to LPC GPIO driver for RT685 DSP
- new inputmux driver
- SOC level definition for RT685 DSP
- RT685 DSP devicetree
- RT685 DSP board definition

VitekST commented on include/zephyr/drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.h:
```diff
@@ -0,0 +1,56 @@
+
+static inline void nxp_rtxxx_adsp_ctl_load_section(
+    const struct device *dev, 
+    const struct nxp_rtxxx_adsp_ctl_image* image, 
```
It essentially can.

VitekST committed:
soc: xtensa: Add HiFi4 DSP of NXP i.MX RT685

Add SoC definitions and a linker file for the HiFi 4 DSP on the NXP i.MX
RT685 microcontroller.

Also modify the West manifest - support depends on a different branch of
the Zephyr NXP HAL, which includes some files that were missing when
compiling for the DSP and omits some files that can't compile for the
DSP.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dts: xtensa: Add DT for HiFi 4 on NXP i.MX RT685

Add a basic device tree for the HiFi 4 DSP on the NXP i.MX RT685
microcontroller.

While no peripherals are instantiated, Zephyr is able to technically run
on this target.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
boards: nxp_adsp_rt685: Enable GPIO and Flexcomm UART

Modify nxp_adsp_rt685's DT to instantiate GPIO peripherals and the 0th
Flexcomm peripheral in UART mode. Also instantiate associated
peripherals - Pinctrl, Clkctl, PINT.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: misc: Add nxp,rtxxx-adsp-ctl

Add the nxp,rtxxx-adsp-ctl driver, modify i.MX RT685's ARM DT to
instantiate it.

Responsibility of this driver is to load code executed by Xtensa-family
cores on NXP i.MX RTxxx microcontrollers and to control their run.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

danieldegrasse commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:
```diff
@@ -42,7 +42,10 @@ config LOG
 
 config NXP_ADSP_SIGN_IMAGE
-	bool
+	bool # FIXME: Defined extra.
```
This line should be removed from this commit

danieldegrasse commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:
```diff
@@ -42,7 +42,10 @@ config LOG
+	bool # FIXME: Defined extra.
 	default n
 
+config NXP_INPUTMUX
```
See comment above about `default y` in the `nxp_inputmux/Kconfig` file- if `default y` is present there then these lines won't be needed.

danieldegrasse commented on dts/xtensa/nxp/nxp_rt685.dtsi:
I'm a bit unclear why this entire file has changed in this commit- it looks like maybe spacing was changed? Either way, this change should not be in this commit. Rather, any spacing fixes needed should be squashed into the original commit making the change that needed the spacing fixed.

danieldegrasse reviewed (commented):
PR is looking a lot better now, thanks for the cleanups :). One thing- some of your commits use language like "This commit does xxx". I would format them closer to how Linux requests you word commits (https://docs.kernel.org/process/submitting-patches.html)

> Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behaviour.

VitekST commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2023 Google LLC.
```
Yes. The file was copied from the RT595 Fusion F1 PR.

danieldegrasse replied:
Understood- I'll go ahead and resolve the similar comments to these then :)

VitekST commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.series:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2023 Google LLC.
```
@danieldegrasse Same argument as above - file was copied from the RT595 Fusion F1 PR.

VitekST commented on drivers/misc/nxp_inputmux/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+config NXP_INPUTMUX
+	bool "NXP i.MX INPUTMUX driver"
+	depends on DT_HAS_NXP_INPUTMUX_ENABLED
```
@danieldegrasse This is set in `soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series`. Should I move it here?

VitekST commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,48 @@
+	default 1000
+
+config DYNAMIC_INTERRUPTS
+	default n
```
@danieldegrasse It does not.

kartben commented on soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright (c) 2023 NXP LLC.
```
FWIW I think the "(c)" is actually pretty common

```
❯ git grep "Copyright [^(]" | wc -l
1861
❯ git grep "Copyright (c)" | wc -l
20743
❯ git grep "Copyright [^(].*NXP"  | wc -l
729
❯ git grep "Copyright (.*NXP"  | wc -l
569
```

kartben commented on boards/arm/mimxrt685_evk/doc/index.rst:
```diff
@@ -327,6 +327,26 @@ steps:
+  ``xcc`` or ``xt-clang``).
+- Build the project with:
+
+.. code-block:: shell
+
+   $ west build -b nxp_adsp_rt685 samples/hello_world
```
use ``.. zephyr-app-commands::``

VitekST committed:
boards: xtensa: Add nxp_adsp_rt685

Add board definitions for the HiFi 4 DSP located on the MIMXRT685-EVK.
Also add a basic board-level DT to get Zephyr to technically run.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: misc: Add nxp,inputmux driver

The responsibility of this driver is to control routing of internal
signals on various NXP microcontrollers.

Importance of this driver in relation to the i.MX RT685 HiFi 4 DSP
support is that it controls routing of interrupt lines from peripherals
to the DSP core.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST commented on boards/arm/mimxrt685_evk/doc/index.rst:
```diff
@@ -327,6 +327,26 @@ steps:
+===============
+
+The Cadence HiFi 4 DSP core is currently supported by Zephyr and Cadence's proprietary toolchain.
+To build a project:
```
@kartben Change committed locally and force pushed in a modified originating commit rather than committed here.

VitekST commented on soc/xtensa/nxp_adsp/common/include/soc.h:
```diff
@@ -30,4 +30,61 @@ extern void z_soc_irq_enable(uint32_t irq);
 extern void z_soc_irq_disable(uint32_t irq);
 extern int z_soc_irq_is_enabled(unsigned int irq);
 
+/* TODO: Reorganize - this is specific for RT5xx/RT6xx */
```
@danieldegrasse Okay, code moved.

andyross reviewed (commented):
Just discovered this in my queue.

Can you clarify the naming here?  "HiFi4" is a coprocessor defined in the Xtensa architecture.  I saw this PR and immediately got excited because I thought it meant "Someone is submitting arch-layer integration for the SIMD instructions".

But in fact this is just a SOC/Board integration for the DSP component of this bigger Cortex-M33 device.   And it happens to be a Xtensa core with HiFi4 instantiated.

I'd just call it "DSP" or "Xtensa CPU" or something?   Note that Zephyr doesn't have proper coprocessor support yet (though there is an integration for the scalar FPU which shares some properties), so you can't actually use the HiFi4 instructions anyway.

andyross commented:
Agreed marketing gets confusing.  All I was really saying was that a PR named "add support or the HiFi4 DSP" sounds like you're adding support for SIMD register state and not just a board. :)

It looks like the naming in the filesystem is "nxp_adsp" though, which is perfectly clear.   No -1 from me.

dbaluta commented:
> Agreed marketing gets confusing. All I was really saying was that a PR named "add support or the HiFi4 DSP" sounds like you're adding support for SIMD register state and not just a board. :)

Agree, coming from @andyross  direction it is a little bit confusing. But this is our like our 5th IMX platform which instantiates HIFI4 DSP and we used this formulation.

We could eventually say: "Add support for enabling HIFI4 DSP instance on MIMXRT685"

danieldegrasse commented on drivers/misc/nxp_rtxxx_adsp_ctl/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config NXP_RTXXX_ADSP_CTL
+	bool "NXP i.MX RTxxx DSP control"
+	depends on DT_HAS_NXP_RTXXX_ADSP_CTL_ENABLED
```
Is the lack of a `default y` here intentional? Usually we add one so all the use needs to do is enable the devicetree node, but I understand if we want to avoid `default y` here since the driver is in the MISC folder (and therefore does not depend on a subystem specific Kconfig)

VitekST replied:
@danieldegrasse It's not actually. It compiles regardless of its inclusion, but I'll add it in.

danieldegrasse commented on dts/bindings/misc/nxp,rtxxx-adsp-ctl.yaml:
```diff
@@ -0,0 +1,31 @@
+  sysctl:
+    required: true
+    type: phandle
+    description: phandle to the SYSCTL (most likely clkctl0) node
```
Why does the comment here state "most likely clkctl0" ?

VitekST replied:
@danieldegrasse That is an artifact of a thought process that was wrong. Will remove.

danieldegrasse commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,45 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
+
+config CACHE
```
Are the `default n` values needed here? IE if we drop them, would these symbols be enabled?

VitekST replied:
@danieldegrasse Yes, they are needed and they will be enabled. It seems like only `NXP_ADSP_SIGN_IMAGE` is redundant, I'll remove that.

danieldegrasse commented on soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc:
```diff
@@ -0,0 +1,14 @@
+	config SOC_NXP_RT685
+		bool "NXP RT685"
+		select HAS_MCUX
+		select HAS_MCUX_FLEXCOMM
```
We don't need to select these Kconfigs to enable the drivers, that is now driven by devicetree. SOCs that select this currently just do so for legacy reasons (and can be cleaned up)

dbaluta commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -429,8 +427,7 @@ static const clock_ip_name_t gpio_clock_names[] = GPIO_CLOCKS;
 		.gpio_base = GPIO,							\
```
@VitekST Lets put this in another PR so that we can merge it faster. 

As per discussion with @stephanosio  the toolchain will be available at the beginning of February.

https://github.com/zephyrproject-rtos/sdk-ng/pull/721#issuecomment-1886648261

VitekST replied:
@dbaluta Okay, I'll create one.

VitekST replied:
@dbaluta Here you go: https://github.com/zephyrproject-rtos/zephyr/pull/67667

danieldegrasse commented on boards/arm/mimxrt685_evk/doc/index.rst:
```diff
@@ -62,52 +62,52 @@ configuration supports the hardware features below.  Another very similar
++-----------+------------+-------------------------------------+------------------+
+| Interface | Controller | Driver/Component                    | Enabled on cores |
++===========+============+=====================================+==================+
+| NVIC      | on-chip    | nested vector interrupt controller  | M4               |
```
Documentation here needs to be fixed- there is no M4 core, the RT685 has an CM33 and HiFi4 DSP

VitekST replied:
I for some reason thought the thing had a Cortex-M3 core. Nevermind, fixed.

dbaluta commented on west.yml:
```diff
@@ -249,7 +249,7 @@ manifest:
     - name: hal_xtensa
-      revision: 08325d6fb7190a105f5382d35e64ed2812c57cf4
+      revision: pull/26/head
       path: modules/hal/xtensa
```
Please update this with the hash of the merged commit or last existing commit in hal_xtensa.

nordicjm reviewed (changes requested):
Needs to be migrated to hwmv2

dbaluta commented:
@VitekST do you have bandwith to look at porting this to hardware model v2?

You should basically follow the similar change here: https://github.com/zephyrproject-rtos/zephyr/pull/69269

VitekST commented:
@dbaluta I don't think I have that much time to spare until I have to turn in my thesis about porting Zephyr and SOF to the target platform, so I would rather focus on the SOF side of things. At the very least, I don't see the point of making the changes until the `collab-hwm` branch is merged.

dbaluta commented:
@VitekST  ok, no problem. I wil keep an eye on this and see if can convert the patches next week.

VitekST committed:
drivers: dma_mcux_lpc: Fix omitted interrupt config

Add back the invocation of an interrupt config function
(config->irq_config_func). Absence of this call results in the DMA
driver not being able to service interrupts raised by the DMA
peripheral.

This case was observed on the i.MX RT685's HiFi 4 DSP domain, where DMA
was not functional because of this.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

hakehuang commented on soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h:
```diff
@@ -0,0 +1,47 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
```
need change to ifndef ...

dbaluta commented:
@VitekST  started on porting this to hwmv2. First patch of the series with some preparation is here: https://github.com/zephyrproject-rtos/zephyr/pull/70952

Will close this PR and send a follow up with support for hifi4.

VitekST commented:
@dbaluta Lovely, thank you.
